contracts/upgradeable_contracts/ChaiConnector.sol

```diff
@@ -0,0 +1,156 @@
+        // the value of chaiBalance() - investedAmountInChai() will be floored,
+        // leading to excess remaining chai balance
+        chaiToken().transfer(recipient, chaiBalance() - investedAmountInChai());
+
+        require(dsrBalance() >= investedAmountInDai());
+    }
+
+    /**
+    * @dev Pays interest, in Chai tokens
+    * @param recipient Account address to receive available interest
+    * @param amount Amount of Chai tokens to transfer
+    */
+    function payInterest(address recipient, uint256 amount) external onlyOwner {
+        // check that the remaining chai balance will be sufficient to cover all invested DAI tokens
+        require(chaiBalance() - amount >= investedAmountInChai());
+        chaiToken().transfer(recipient, amount);
+
+        require(dsrBalance() >= investedAmountInDai());
+    }
+
+    /**
+    * @dev Returns current used chai contract
+    * @return chai contract address
+    */
+    function chaiToken() public view returns (IChai) {
+        return IChai(addressStorage[CHAI_TOKEN]);
+    }
+
+    /**
+    * @dev Evaluates bridge balance for tokens, holded in DSR
+    * @return Balance in dai, truncated
+    */
+    function dsrBalance() public view returns (uint256) {
+        uint256 chi = chaiToken().pot().chi();
+        return rmul(chi, chaiBalance());
+    }
+
+    /**
+    * @dev Evaluates bridge balance in Chai tokens
+    * @return Balance in chai, exact
+    */
+    function chaiBalance() public view returns (uint256) {
+        return chaiToken().balanceOf(address(this));
+    }
+
+    /**
+    * @dev Evaluates exact current invested amount, id DAI
+    * @return Value in DAI
+    */
+    function investedAmountInDai() public view returns (uint256) {
+        return uintStorage[INVESTED_AMOUNT];
+    }
+
+    function setInvestedAmointInDai(uint256 amount) internal {
+        uintStorage[INVESTED_AMOUNT] = amount;
+    }
+
+    /**
+    * @dev Evaluates amount of chai tokens that is sufficent to cover 100% of the invested DAI
+    * @return Amount in chai, ceiled
+    */
+    function investedAmountInChai() internal returns (uint256) {
+        IPot pot = chaiToken().pot();
+        // solhint-disable-next-line not-rely-on-time
+        uint256 chi = (now > pot.rho()) ? pot.drip() : pot.chi();
+        return rdivup(investedAmountInDai(), chi);
+    }
+
+    /**
+    * @dev Invests DAI into Chai
+    * @param amount Amount of DAI to invest
+    */
+    function _convertDaiToChai(uint256 amount) internal {
+        setInvestedAmointInDai(investedAmountInDai() + amount);
+        erc20token().approve(chaiToken(), amount);
+        chaiToken().join(address(this), amount);
+    }
+
+    /**
+    * @dev Redeems DAI from Chai, the total redeemed amount will be at least equal to specified amount
+    * @param amount Amount of DAI to redeem
+    */
+    function _convertChaiToDai(uint256 amount) internal {
+        uint256 invested = investedAmountInDai();
+        if (amount >= invested) {
```

Comment:
What is the case when `amount >= invested`? Could this appear as a valid scenario?

Reply:
The only case is related to the `convertChaiToDai(uint256 _amount)`, so that it will just convert all remaining chai into dai instead of revert on too high `_amount`.

Reply:
There is another case, by the way. In `onExecuteMessage` bridge will try to withdraw tokens, so that the balance will be equal to `minDaiTokenBalance()` after the operation. If the overall bridge balance is less than this limit, it will just withdraw all remaining tokens.

Reply:
> The only case is related to the convertChaiToDai(uint256 _amount), so that it will just convert all remaining chai into dai instead of revert on too high _amount.

From my perspective it is better to revert by requiring explicit actions from the owners. But I would insist of removing `convertChaiToDai(amount)` at all.

Reply:
> If the overall bridge balance is less than this limit, it will just withdraw all remaining tokens.

Got you! Please add the corresponding comment to the code to not forget this case. Is this scenario covered in tests?

Reply:
Not properly. I will add an explicit test on that.